### PR TITLE
ALMA: auth handling to login to Keycloak

### PR DESCRIPTION
**Pull request by at88mph**
Added auth handling to login to Keycloak.

Fixes #2670

### CODE REVIEW

**Comment by keflavich** [comment]
@andamian could you provide the review for this one?  It looks good to me but I haven't tested and I know nothing about the ALMA login protocol.

File: astroquery/alma/core.py [hunk 1]
```diff
@@ -212,6 +212,91 @@ def _gen_sql(payload):
     return sql + where
 
 
+#
+# Authentication session information for connecting to an OIDC instance.  Assumes an OIDC system like Keycloak
+# with a preconfigured client called "oidc".
```
Comment (andamian) [inline astroquery/alma/core.py:217]: In spite of the name and the description, this class only checks that the user/password is valid with any of the Keycloak systems deployed in a list of URLs. Nothing else from the key cloak functionality (such as tokens) are further used in the scheme (but maybe it should?). 

This confused me a great deal as I was trying to figure out what kind of magic Keycloak employs to facilitate authentication on the services so I suggest adding the above paragraph or something to that extend to make it clear and prevent future confusion.

File: astroquery/alma/core.py [hunk 2]
```diff
@@ -212,6 +212,91 @@ def _gen_sql(payload):
...
+    _REALM_ENDPOINT = '/auth/realms/ALMA'
+    _LOGIN_ENDPOINT = f'{_REALM_ENDPOINT}/protocol/openid-connect/token'
+    _VERIFY_WELL_KNOWN_ENDPOINT = f'{_REALM_ENDPOINT}/.well-known/openid-configuration'
+
+    def __init__(self):
+        super().__init__()
+        self._auth_hosts = auth_urls
+        self._auth_host = None
+
+    def set_auth_hosts(self, auth_hosts):
```
Comment (andamian) [inline astroquery/alma/core.py:232]: I would make this a property setter and fail on `None`. This way it gets invoked in the ctor too.

File: astroquery/alma/core.py [hunk 3]
```diff
@@ -212,6 +212,91 @@ def _gen_sql(payload):
...
+        auth_hosts : array
+            Available hosts name.  Checking each one until one returns a 200 for
+            the well-known endpoint.
+        """
+        self._auth_hosts = auth_hosts
+
+    @property
+    def host(self):
+        if self._auth_host is None:
+            for auth_url in self._auth_hosts:
```
Comment (andamian) [inline astroquery/alma/core.py:247]: Currently this returns a generic `TypeError: 'NoneType' object is not iterable` if `self._auth_hosts` is not set properly. See my comment above.

File: astroquery/alma/core.py [hunk 4]
```diff
@@ -212,6 +212,91 @@ def _gen_sql(payload):
...
+        json_auth = login_response.json()
+
+        if 'error' in json_auth:
+            log.debug(f'{json_auth}')
+            error_message = json_auth['error_description']
+            if self._INVALID_PASSWORD_MESSAGE not in error_message:
+                raise LoginError("Could not log in to ALMA authorization portal: "
+                                 f"{self.host} Message from server: {error_message}")
+            else:
+                log.error(error_message)
```
Comment (andamian) [inline astroquery/alma/core.py:292]: Missing `LoginError`. I think it should always raise an exception unless it succeeds.

File: astroquery/alma/core.py [hunk 5]
```diff
@@ -212,6 +212,91 @@ def _gen_sql(payload):
...
+        Authenticate to one of the configured hosts.
+
+        Parameters
+        ----------
+        username : str
+            The username to authenticate with
+        password : str
+            The user's password
+        """
+        data = {
```
[the remark this answers is not in the document]
Comment (at88mph) [reply astroquery/alma/core.py:273]: That might be true.  If the `self._auth_host` is set, that just means that `self.get_valid_host()` was called at some point, not that the current user is logged in.

[the remark this answers is not in the document]
Comment (andamian) [reply astroquery/alma/core.py:283]: Right. Please ignore my comment.

**Review by andamian — APPROVED** [review]

**Comment by andamian** [comment]
@bsipocz - FYI: this is ready from our perspective.

**Review by keflavich — APPROVED** [review]

**Comment by bsipocz** [comment]
Thank you @at88mph and @andamian! I see two, genuine-looking alma failures, but both are unrelated and are present on `main`, too.

**Pull request merged**